redis-utils-lib/src/cmr/redis_utils/redis.clj

```diff
@@ -6,39 +6,52 @@
    [taoensso.carmine :as carmine]))
 
-(defmacro wcar*
-  "Safe call redis with conn opts and retries.
-  For available commands refer to Redis docuemntation or:
-  https://github.com/ptaoussanis/carmine/blob/master/src/taoensso/carmine/commands.edn."
-  [& body]
-  `(let [with-retry#
-         (fn with-retry# [num-retries#]
-             (try
-               (carmine/wcar (config/redis-conn-opts) ~@body)
-               (catch Exception e#
-                 (if (> num-retries# 0)
-                   (do
-                     (info (format "Redis failed with exception %s. Retrying %d more times."
-                                   e#
-                                   (dec num-retries#)))
-                     (Thread/sleep (config/redis-retry-interval))
-                     (with-retry# (dec num-retries#)))
-                   (error "Redis failed with exception " e#)))))]
-     (with-retry# (config/redis-num-retries))))
+  (defmacro wcar*
+    "Safe call redis with conn opts and retries.
+    The passed in key allows for specific redis caches as well as more logging information.
+    The read? is a true/false flag that signifies that a read-only replica redis connection
+    should be used. This flag is also used in more logging information.
+    For available commands refer to Redis docuemntation or:
+    https://github.com/ptaoussanis/carmine/blob/master/src/taoensso/carmine/commands.edn."
+    [key read? & body]
+    `(let [with-retry#
+            (fn with-retry# [num-retries#]
+              (let [wr-string# (if ~read? "read" "write")
+                    config# (if (= :collection-metadata-cache ~key)
+                              (if ~read?
+                                (config/redis-collection-metadata-cache-read-conn-ops)
+                                (config/redis-collection-metadata-cache-conn-ops))
+                              (if ~read?
+                                (config/redis-read-conn-ops)
+                                (config/redis-conn-opts)))]
+                (try
+                  (carmine/wcar config# ~@body)
+                  (catch Exception e#
+                    (if (> num-retries# 0)
+                      (do
+                        (info (format "Redis %s %s failed with exception %s. Retrying %d more times."
+                                      wr-string#
+                                      ~key
+                                      e#
+                                      (dec num-retries#)))
+                        (Thread/sleep (config/redis-retry-interval))
+                        (with-retry# (dec num-retries#)))
+                      (error (format "Redis %s %s failed with exception %s"  wr-string# ~key e#)))))))]
+        (with-retry# (config/redis-num-retries))))
 
 (defn healthy?
   "Returns true if able to reach Redis."
   [& args]
   {:ok?
    (try
-     (= "PONG" (wcar* (carmine/ping)))
+     (= "PONG" (wcar* "healthy?" true (carmine/ping)))
```

Comment:
Do we not need to use this function on the redis-collection-metadata-cache? With new code for wcar* it seems like 'healthy?' 'reset' and 'get-keys' will never be able to be used on the new redis cache